include/rocksdb/file_system.h

```diff
@@ -0,0 +1,1356 @@
+struct IODebugContext {
+  // file_path to be filled in by RocksDB in case of an error
+  std::string file_path;
+
+  // A map of counter names to values - set by the FileSystem implementation
+  std::map<std::string, uint64_t> counters;
+
+  // To be set by the FileSystem implementation
+  std::string msg;
+
+  IODebugContext() {}
+
+  void AddCounter(std::string& name, uint64_t value) {
+    counters.emplace(name, value);
+  }
+
+  std::string ToString() {
+    std::ostringstream ss;
+    ss << file_path << ", ";
+    for (auto counter : counters) {
+      ss << counter.first << " = " << counter.second << ",";
+    }
+    ss << msg;
+    return ss.str();
+  }
+};
+
+// The FileSystem, FSSequentialFile, FSRandomAccessFile, FSWritableFile,
+// FSRandomRWFileclass, and FSDIrectory classes define the interface between
+// RocksDB and storage systems, such as Posix filesystems,
+// remote filesystems etc.
+// The interface allows for fine grained control of individual IO operations,
+// such as setting a timeout, prioritization, hints on data placement,
+// different handling based on type of IO etc.
+// This is accomplished by passing an instance of IOOptions to every
+// API call that can potentially perform IO. Additionally, each such API is
+// passed a pointer to a IODebugContext structure that can be used by the
+// storage system to include troubleshooting information. The return values
+// of the APIs is of type IOStatus, which can indicate an error code/sub-code,
+// as well as metadata about the error such as its scope and whether its
+// retryable.
+class FileSystem {
+ public:
+  FileSystem();
+
+  // No copying allowed
+  FileSystem(const FileSystem&) = delete;
+
+  virtual ~FileSystem();
+
+  virtual const char* Name() const = 0;
+
+  static const char* Type() { return "FileSystem"; }
+
+  // Loads the environment specified by the input value into the result
+  static Status Load(const std::string& value,
+                     std::shared_ptr<FileSystem>* result);
+
+  // Return a default fie_system suitable for the current operating
+  // system.  Sophisticated users may wish to provide their own Env
+  // implementation instead of relying on this default file_system
+  //
+  // The result of Default() belongs to rocksdb and must never be deleted.
+  static std::shared_ptr<FileSystem> Default();
+
+  // Create a brand new sequentially-readable file with the specified name.
+  // On success, stores a pointer to the new file in *result and returns OK.
+  // On failure stores nullptr in *result and returns non-OK.  If the file does
+  // not exist, returns a non-OK status.
+  //
+  // The returned file will only be accessed by one thread at a time.
+  virtual IOStatus NewSequentialFile(const std::string& fname,
+                                     const FileOptions& file_opts,
+                                     std::unique_ptr<FSSequentialFile>* result,
+                                     IODebugContext* dbg) = 0;
```

Comment:
> * Why is IODebugConter rather than a std::shared_ptr<>?
> * Why do you need it at all? Why can't it be part of FileOptions?

I think I have addressed this before. The requirement is to be able to trace individual calls to the file system. The file system calls can be sampled, logged and then aggregated (or any other ad-hoc analysis) outside the IO path.

> * If it is not an output parameter, it should not be last

It is an output parameter.

Reply:
Seeing no examples of where this is being used, I still do not understand when you will trace one specific call for one specific file independent of other calls related to that file.  I can understand the use case of having a IODebugContext that is a parameter to the constructor of a file, but not saying that we are tracing only reads for a specific file.  I do not understand how the calling code would differentiate and pass a different DebugContext for file 1 versus file 2 for each of these calls.  

Is it an output parameter (as in something that is created) or an in-out parameter (as in something that might be changed -- more like a Logger class)?

Reply:
Its an in-out parameter. We follow the Google C++ style guide, which recommends placing input-only parameters before output parameters.